tun/checksum.go

```diff
@@ -0,0 +1,42 @@
+package tun
+
+import "encoding/binary"
+
+// TODO: Explore SIMD and/or other assembly optimizations.
+func checksumNoFold(b []byte, initial uint64) uint64 {
+	ac := initial
+	i := 0
+	n := len(b)
+	for n >= 4 {
+		ac += uint64(binary.BigEndian.Uint32(b[i : i+4]))
+		n -= 4
+		i += 4
+	}
+	for n >= 2 {
+		ac += uint64(binary.BigEndian.Uint16(b[i : i+2]))
+		n -= 2
+		i += 2
+	}
+	if n == 1 {
+		ac += uint64(b[i]) << 8
+	}
+	return ac
+}
+
+func checksum(b []byte, initial uint64) uint16 {
+	ac := checksumNoFold(b, initial)
+	for (ac >> 16) > 0 {
+		// TODO: Test fixed number of folds instead of branching.
```

Comment:
Probably a good TODO to take care of now, right? This operates on plaintext, so we probably don't want non-constant time things happening?

Reply:
[Done.](https://github.com/WireGuard/wireguard-go/pull/64/commits/1e450d5da977d5325f1f44c266568d7a82f54a9f)